bbot/modules/internal/excavate.py

```diff
@@ -1,476 +1,858 @@
-        async for result, name in self._search(content, event, **kwargs):
-            results.add((result, name))
-        for result, name in results:
-            await self.report(result, name, event, **kwargs)
-
-    async def _search(self, content, event, **kwargs):
-        for name, regex in self.compiled_regexes.items():
-            # yield to event loop
-            await self.excavate.helpers.sleep(0)
-            for result in await self.helpers.re.findall(regex, content):
-                yield result, name
-
-    async def report(self, result, name, event):
-        pass
-
-
-class CSPExtractor(BaseExtractor):
-    regexes = {"CSP": r"(?i)(?m)Content-Security-Policy:.+$"}
-
-    async def extract_domains(self, csp):
-        domains = await self.helpers.re.findall(dns_name_regex, csp)
-        unique_domains = set(domains)
-        return unique_domains
-
-    async def search(self, content, event, **kwargs):
-        async for csp, name in self._search(content, event, **kwargs):
-            extracted_domains = await self.extract_domains(csp)
-            for domain in extracted_domains:
-                await self.report(domain, event, **kwargs)
-
-    async def report(self, domain, event, **kwargs):
-        await self.excavate.emit_event(
-            domain,
-            "DNS_NAME",
-            parent=event,
-            tags=["affiliate"],
-            context=f"{{module}}'s CSP extractor found {{event.type}}: {{event.data}} in CSP header",
-        )
-
-
-class HostnameExtractor(BaseExtractor):
-    regexes = {}
+        self.name = ""
+
+    async def preprocess(self, r, event, discovery_context):
+        self.discovery_context = discovery_context
+
+        description = ""
+        tags = []
+        emit_match = False
+
+        if "description" in r.meta.keys():
+            description = r.meta["description"]
+        if "tags" in r.meta.keys():
+            tags = parse_list_string(r.meta["tags"])
+        if "emit_match" in r.meta.keys():
+            emit_match = True
+
+        yara_rule_settings = YaraRuleSettings(description, tags, emit_match)
+        yara_results = {}
+        for h in r.strings:
+            yara_results[h.identifier.lstrip("$")] = sorted(set([i.matched_data.decode("utf-8") for i in h.instances]))
+        await self.process(yara_results, event, yara_rule_settings)
+
+    async def process(self, yara_results, event, yara_rule_settings):
+
+        for identifier, results in yara_results.items():
+            for result in results:
+                event_data = {"host": str(event.host), "url": event.data.get("url", "")}
+                event_data["description"] = f"{self.discovery_context} {yara_rule_settings.description}"
+                if yara_rule_settings.emit_match:
+                    event_data["description"] += f" [{result}]"
+                await self.report(event_data, event, yara_rule_settings)
+
+    async def report_prep(self, event_data, event_type, event, tags):
```

Comment:
Each of the functions on the base `ExcavateRule` class need to have docstrings that explain their purpose and how they fit with the other functions.

Reply:
https://github.com/blacklanternsecurity/bbot/pull/1471/commits/7560cc8fc393d58fb397aad2b13ff83d51745d60
https://github.com/blacklanternsecurity/bbot/pull/1471/commits/5ded168f39c61e284ecb2620ad0b2cfe6aa300b2